### ISSUE TITLE
Alternation syntax for destructuring

### ISSUE THREAD

**Issue by nicowilliams** [stedolan/jq#856]
try/catch is nice, but the `?` shorthand for `try EXP catch empty` is very convenient.  It'd be nice if there was a `?||` binary operator that caused the second expression to be evaluated IFF (or, rather, _when_) the first produces an error.  I.e., something like `EXP ?|| EXP`, and then destructuring should also use this, so that one could write `. as {first:$first}?|| [$first] | ...`.

**Comment by dtolnay** [comment]
How does this apply to destructuring?

**Comment by nicowilliams** [comment]
$ ./jq -n '{first:1}|. as [$first] | $first'
    jq: error (at <unknown>): Cannot index object with number

The `?` syntactic sugar for handling errors is quite convenient, but not convenient enough.

**Comment by dtolnay** [comment]
Got it. Thanks for clarifying the top comment.

**Comment by nicowilliams** [comment]
So, for example, if `.` could match one of two different schemas, an error alternation syntactic sugar would make it easy to write a single destructuring expression for it.  Basically, it's alternation.  To be sure, we don't really have a way to require matches either, so that `. as {$first, $second} | [$first, $second]` outputs `[null, null]` when the first and second keys are missing in `.`, so the problem here is a bit more general.

But even ignoring destructuring, a short-hand for `. as $dot | try EXP catch ($dot | ALTEXP) | ...` would be convenient.  I'm thinking that `EXP ?|| ALTEXP` would be that.

I considered `?:`, but I think that would be confusing since it would resemble trinary operators elsewhere, but would only take the else branch when the condition raises an error, rather than when the condition is false, and maybe we should reserve the ability to add a proper trinary operator later.

**Comment by nicowilliams** [comment]
The Icon programming language has a backslash unary operator for "is not null", which might be worth copying, though it's not obvious (unlike `?`, which is), so maybe not.

**Comment by dtolnay** [comment]
`EXP ?// EXP` basically already does what you want. Using it for destructuring is not perfect:

    david@genie:/github/jq$ ./jq -n '{first:1} | (. as {first:$first} | $first) ?// (. as [$first] | $first) | . + 1'
    2
    david@genie:/github/jq$ ./jq -n '[1] | (. as {first:$first} | $first) ?// (. as [$first] | $first) | . + 1'
    2

So my counterproposal is `. as {first:$first} ?// [$first] | $first`.

**Comment by dtolnay** [comment]
Never mind, the semantics are different when the first pattern matches but $first is false or null. `?||` is better.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> . as $dot | try EXP catch ($dot | ALTEXP) 

Perhaps I'm missing something but I believe there's already a (satisfactory) shorthand:
```
try (EXP) // ALTEXP 
```
Here are some examples:
```
$ jq -n '1 | . as $dot | try (. / 0) // $dot'
1

$ jq -n '1 | . as $dot | try (. / 0) // .'
1

$ jq -n '1 | try (. / 0) // .'
1
```

**Comment by nicowilliams** [comment]
@pkoppstein Oh, hmm, indeed, `//` does eval the RHS when the LHS produces no values.  So `? //` actually works as I'd like.  This leaves only something for destructuring.

As for destructuring, there is a complication: every binding on the LHS and RHS of `?//` has to be established.

**Comment by nicowilliams** [comment]
I've changed the issue's title since `?//` works today.

**Comment by nicowilliams** [comment]
We probably don't need the `?` in the destructuring syntax, though if we keep it then `?//` to match the non-destructuring case seems best, otherwise I'd say just `||` instead of `?||`.

**Comment by dtolnay** [comment]
I don't like `||` because it is too similar to `|` which ends the destructuring.

    . as [$first] || {first:$first} || $fallback | ...

Just to brainstorm some ideas:

    . as [$first] or {first:$first} or $fallback | ...
    . as try [$first], {first:$first}, $fallback | ...
    . as try [$first], try {first:$first}, $fallback | ...
    . as try ([$first], {first:$first}, $fallback) | ...
    . as [$first]?, {first:$first}?, $fallback | ...
    . as [$first]? // {first:$first}? // $fallback | ...
    . as [$first] // {first:$first} // $fallback | ...

**Comment by pkoppstein** [comment]
Yes, using "||" here would be error-prone and perhaps confusing.  The first and last options ("or" and "//") are nice, not least because they read well.  " _ as _ or _ or _" is particularly clear. The others by comparison seem somewhat clumsy and/or verbose and/or potentially confusing.

Another alternative that would be even sleeker is:
```
. as [$first] ! {first:$first} ! $fallback | ..
```

**Comment by nicowilliams** [comment]
`or` and `//` work for me.

I think we must insist that either all alternatives establish the same bindings, or that all bindings are established with `null` if not not matched.  E.g., `. as {$foo} or $bar | ...` must bind both, `$foo` *and* `$bar`, and must leave one of the two as null (possibly both, of course).

@pkoppstein I want to leave `!` available for use as a sort of opposite of `?` (see #859).

**Comment by dtolnay** [comment]
I like `or`, but do either of you have a real-life example where this feature would be useful? I would like to make sure we are solving the use case in the best way before adding complexity to the language. It could be that anywhere this would be useful would be better served by a pattern matching switch statement or something else entirely.

**Comment by nicowilliams** [comment]
@dtolnay If a schema permits something to be either an object or an array...  For example, consider what we might call ASN.1 JSON Encoding Rules.  There are two obvious way to encode an ASN.1 `SEQUENCE { ... }` in JSON: as an array (whose elements correspond to the SEQUENCE's fields in the definition order, with absent OPTIONAL fields encoded as `null`), or as an object (with no particular field order, and absent fields... absent).  The latter is more convenient to anyone reading the encoded text, but the former is more compact.  One could easily decide that both are permitted.  Then one could parse a `SEQUENCE { foo INTEGER, bar INTEGER }` as `. as {$foo, $bar}` or [$foo, $bar] | ...`.  (There is a duality to arrays and objects in that either can be used to represent rigid structures.  When a hash can contain arbitrary keys, objects become the superior choice.)

I think #859 is much more interesting in the short term.

These are really not urgent issues.  I made these 1.6 milestones, not 1.5.  The point of filing these now is to not forget them.

**Comment by pkoppstein** [comment]
@dtolnay wrote:
> It could be that anywhere this would be useful would be better served by a pattern matching switch statement or something else entirely.

Excellent point!  jq sorely needs a case/when construct, if only because at present one often has to obfuscate one's code with numerous occurrences of "$dot" or equivalent.  For example, to switch on 'type', one either has to repeat "type" or write something like:
```
. as $dot | type | if . == "number" then $dot | NUMBER elif . == "boolean" then $dot | BOOLEAN .... end
```
It would be so much better to be able to write something like:
```
case type
  when "number" NUMBER
  when "boolean" BOOLEAN
end
```  

A related point (in favor of focusing on case/when rather than fancy destructuring in the near term) is that jq is currently lacking some of the infrastructure needed to make pattern-matching and destructuring all that one would like.  For pattern-matching with arrays, one would like to be able to specify a head/tail pattern (as in Prolog or Elixir), but currently one cannot write [$head | $tail] or [$head, $tail], at least not without creating confusion and/or inconsistencies.  

One possibility would be to allow [$head & $tail].  

In summary, it seems to me that even a relatively simple case/when construct is greatly to be wished. If it is designed with an eye to the future, I even dare to hope it might make it into 1.5 :-)

**Comment by nicowilliams** [comment]
This is actually done, though there's a bug that needs fixing.
